File: 1020-number-of-enclaves/1020-number-of-enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        
        rows, cols = len(grid), len(grid[0])
        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == 0:
                return 0
            grid[r][c] = 0
            return 1 + dfs(r+1, c) + dfs(r-1, c) + dfs(r, c-1) + dfs(r, c+1)
         
        for r in range(rows):
            for c in range(cols):
                if (r == 0 or c == 0 or r == rows - 1 or c == cols - 1) and grid[r][c] == 1:
                    dfs(r,c)
        return sum(sum(row) for row in grid)

File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    
    int dfs(vector <vector <int>>& grid, int r, int c)
    {
        if(r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == 0)
            return 0;
        grid[r][c] = 0;
        return 1 + dfs(grid, r + 1, c) + dfs(grid, r - 1 , c) + dfs(grid, r, c + 1) + dfs(grid, r, c - 1);
    }
    
    int numEnclaves(vector<vector<int>>& grid) 
    {
        int rows = grid.size(); int cols = grid[0].size();
        for(int r = 0; r < rows; r++)
        {
            for(int c = 0; c < cols; c++)
            {
                if((r == 0 || r == rows - 1 || c == 0 || c == cols - 1) && grid[r][c] == 1)
                    dfs(grid, r, c);
            }
        }
        int enclaves = 0;
        for(int r = 0; r < rows; r++)
        {
            for(int c = 0; c < cols; c++)
            {
                if(grid[r][c] == 1)
                    enclaves += dfs(grid, r, c);
            }
        }
        
        return enclaves;
    }
};